### PR TITLE
More intelligent transaction requests

### LINKED ISSUE #1161: More intelligent transaction requests
**UPDATE:** new plan described in https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-594423279

Old plan:

---

Currently the `txSubmissionInbound` protocol handler does not look at the Mempool to see if txs are already there before requesting them. So in practice, it will needlessly download txs it already has. More confusingly, we can observe that node A sends a tx to B, which then sends it right back to A.

There are multiple options to improve this situation:
1. Look at the Mempool and don't request txs already in there.
2. Maintain a shared (across all `txSubmissionInbound` protocol handlers) set of `TxId`s that were recently in the Mempool. They should also expire from this set, so perhaps a PSQ would be a good fit. This can subsume (1).
3. Keep track of which `TxId`s we're asking from which peers and don't ask for the same one from multiple peers unless that would otherwise cause things to stall, in which case we should make another request for the same `TxId`.

Because of the more complex concurrency of (3), we'll keep that for later. Getting (1) working will be easy. Our current aim is to achieve (2).

### PR DESCRIPTION
**Pull request by intricate**
Closes #1161.

Implements the [new plan](https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-594423279).

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 1]
```diff
@@ -216,6 +232,53 @@ data Mempool m blk idx = Mempool {
...
+isTxAdded TxAdded = True
+isTxAdded _       = False
+
+isTxRejected :: MempoolAddTxResult blk -> Bool
+isTxRejected (TxRejected _) = True
+isTxRejected _              = False
+
+-- | An error encountered when attempting to add a transaction to the mempool.
+data MempoolAddTxError blk
+  = TxApplicationError !(ApplyTxErr blk)
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]: Is the word "application" misleading?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 2]
```diff
@@ -216,6 +232,53 @@ data Mempool m blk idx = Mempool {
     , zeroIdx        :: idx
     }
 
+-- | The result of attempting to add a transaction to the mempool.
+data MempoolAddTxResult blk
+  = TxAdded
+    -- ^ The transaction was added to the mempool.
+  | TxRejected !(MempoolAddTxError blk)
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]: Should these constructors be prefixed with `Mempool`?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 3]
```diff
@@ -472,6 +522,11 @@ internalStateFromVR vr = IS {
       , vrLastTicketNo
       } = vr
 
+    txIdSet = foldl'
+      (\acc tx -> Set.insert (txId tx) acc)
+      Set.empty
+      vrValid
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: I could implement this more efficiently by adding a new field to `ValidationResult` and appending to it whenever `vrValid` is appended to.

**Comment by nfrisby** [comment]
I rebased my wip branch onto commit 8cd528bb, and I no longer see the huge number of transactions. Great!

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs (hunk 3 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Yeah, that's a good idea, it brings `ValidationResult` and `InternalState` more in sync again.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs [hunk 4]
```diff
@@ -252,13 +292,61 @@ txSubmissionInbound _tracer maxUnacked TxSubmissionMempoolWriter{..} =
...
+    -- are already in the mempool and return the updated 'ServerState'.
+    --
+    -- This enables us to effectively filter out transactions that we don't
+    -- need to bother requesting from the client since they're already in the
+    -- mempool.
+    --
+    acknowledgeTxIdsInMempool :: ServerState txid tx
+                              -> MempoolSnapshot txid tx idx
+                              -> ServerState txid tx
+    acknowledgeTxIdsInMempool st MempoolSnapshot{mempoolHasTx} = do
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs]: This function is not-monadic, so I wouldn't use `do` here.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs [hunk 5]
```diff
@@ -192,16 +226,22 @@ txSubmissionInbound _tracer maxUnacked TxSubmissionMempoolWriter{..} =
...
+      -- transactions again in the future.
+      let st' = st {
         requestedTxIdsInFlight = requestedTxIdsInFlight st - reqNo,
         unacknowledgedTxIds    = unacknowledgedTxIds st
                               <> Seq.fromList (map fst txids),
         availableTxids         = availableTxids st
                               <> Map.fromList txids
       }
+      mpSnapshot <- atomically mempoolGetSnapshot
+      serverIdle n (acknowledgeTxIdsInMempool st' mpSnapshot)
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs]: I'm sure you and @dcoutts have discussed this already, but just checking again: is this the right and *only* place to look at the Mempool?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]: Hmm, for consistency, yeah.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 6]
```diff
@@ -216,6 +232,53 @@ data Mempool m blk idx = Mempool {
...
+isTxAdded :: MempoolAddTxResult blk -> Bool
+isTxAdded TxAdded = True
+isTxAdded _       = False
+
+isTxRejected :: MempoolAddTxResult blk -> Bool
+isTxRejected (TxRejected _) = True
+isTxRejected _              = False
+
+-- | An error encountered when attempting to add a transaction to the mempool.
+data MempoolAddTxError blk
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]: The type Edsko and I suggested in https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-594423279 intentionally did not classify "already in Mempool" as an error, as it's not really an error. For [example](https://github.com/input-output-hk/ouroboros-network/issues/1161#issuecomment-595093203), the wallet will resubmit transactions that didn't make it into the last block via the `LocalTxSubmissionServer`. The wallet doesn't know that the transaction is still in the Mempool, so we shouldn't treat that case as an error (think of it as a trace message severity). So go for something like:

```haskell
data MempoolAddTxResult blk
  = MempoolTxAdded
  | MempoolTxRejected (ApplyTxErr blk)
  | MempoolTxAlreadyInMempool
```

Also important: this new `MempoolAddTxError` type replaces the existing `ApplyTxError` type in the `LocalTxSubmission` protocol codec, so this breaks network compatibility, which we have to plan carefully. The 3-constructor approach doesn't break network compatibility.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalTxSubmission/Server.hs [hunk 7]
```diff
@@ -32,9 +32,11 @@ localTxSubmissionServer tracer mempool =
         traceWith tracer $ TraceReceivedTx tx
         res <- addTxs mempool [tx]
         case res of
-          [(_tx, mbErr)] -> return (mbErr, server)
+          [(_tx, addTxRes)] -> case addTxRes of
+            TxAdded             -> return (Nothing, server)
+            TxRejected addTxErr -> return (Just addTxErr, server)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalTxSubmission/Server.hs]: So `MempoolTxAlreadyInMempool` would also result in `(Nothing, server)`.

File: ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs (hunk 5 above)
Comment (intricate) [reply ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs]: In my discussions with Duncan, he noted that this seemed like the right place to do it, yeah. I also believe this is where it makes the most sense as well.

**Pull request merged**